File: estructuras/RTree.cpp
#include "RTree.hpp"
#include <iostream>
#include <math.h>
#include <fstream>
#include <string>
#include <algorithm>
#include <vector>
#include "../algoritmos/RTreeAlgorithm.hpp"
#include "../disk_manipulation_functions/binaryFileFunctions.hpp"

/**
 * @brief Constructor de la clase RTree.
 * \param maxNodeCapacity Cantidad máxima de hijos por nodo.
 * \param algorithm Algoritmo a utilizar para construir el RTree.
 * \param numberOfRects Cantidad de rectángulos que componen el RTree.
*/
RTree::RTree(int maxNodeCapacity, RTreeAlgorithm* algorithm, unsigned int numberOfRects) {
  this->maxNodeCapacity = maxNodeCapacity;
  this->algorithm = algorithm;
  this->numberOfRects = numberOfRects;
  this->totalSearchIOs = 0;
}

/**
 * @brief Método que retorna el contador de I/Os realizadas en la última búsqueda.
*/
unsigned int RTree::getTotalSearchIOs() {
  return this->totalSearchIOs;
}

/**
 * @brief Método que resetea el contador de I/Os realizadas en la última búsqueda.
*/
void RTree::resetTotalSearchIOs() {
  this->totalSearchIOs = 0;
}

/**
 * @brief Método que construye un RTree a partir de un archivo de rectángulos con un algoritmo.
 * \param filename Nombre del archivo de rectángulos.
*/
void RTree::buildTreeFromFile(std::string filename) {
  this->algorithm->buildTree(filename);
  this->treeFile = &this->algorithm->getTreeFile();
  this->leavesFile = &this->algorithm->getLeavesFile();
}

/**
 * @brief Método que verifica si 2 rectangulos intersecan.
 * \param queryRegion Rectángulo que representa la query.
 * \param MBR Rectángulo que representa el MBR de un nodo o el rectángulo de una hoja.
*/
bool RTree::intersects(Rect queryRegion, Rect MBR) {
  bool xOverlap = queryRegion.x2 >= MBR.x1 && queryRegion.x1 <= MBR.x2;
  bool yOverlap = queryRegion.y2 >= MBR.y1 && queryRegion.y1 <= MBR.y2;

  return xOverlap && yOverlap;
}

/**
 * @brief Método que realiza una búsqueda en el RTree.
 * \param region Rectángulo que representa la query.
*/
std::vector<Rect> RTree::search(Rect region) {
  std::vector<Rect> result;
  this->searchRecursive(region, 0, 0, result);
  return result;
}

/**
 * @brief Método que realiza una búsqueda en el RTree de forma recursiva.
 * \param region Rectángulo que representa la query.
 * \param firstChildIndex Índice del primer hijo del nodo, negativo o 0 si el hijo es hoja.
 * \param lastChildIndex Índice del último hijo del nodo, negativo o 0 si el hijo es hoja.
 * \param result Vector que contiene los rectángulos que intersectan con la query.
*/
void RTree::searchRecursive(Rect region, long long firstChildIndex, long long lastChildIndex, std::vector<Rect>& result) {
  // Si los hijos son hojas
  if (firstChildIndex < 0 || lastChildIndex < 0) {
    long long first = -firstChildIndex / sizeof(Rect);
    long long last = -lastChildIndex / sizeof(Rect);
    long long buffSize = last - first + 1;
    Rect* rectBuffer = new Rect[buffSize];
    std::istream& leavesFile = *this->leavesFile;
    leavesFile.seekg(firstChildIndex, std::ios::beg);
    leavesFile.read(reinterpret_cast<char*>(rectBuffer), sizeof(Rect) * buffSize);
    this->totalSearchIOs += (unsigned int)std::ceil((double)buffSize / rectanglesPerBlock);
    for (long long i = 0; i < buffSize; i++) {
      if (RTree::intersects(region, rectBuffer[i])) {
        result.push_back(rectBuffer[i]);
      }
    }
    delete[] rectBuffer;
  } else {
    long long first = firstChildIndex / sizeof(RTreeNode);
    long long last = lastChildIndex / sizeof(RTreeNode);
    long long buffSize = last - first + 1;
    RTreeNode* nodeBuffer = new RTreeNode[buffSize];
    std::istream& treeFile = *this->treeFile;
    treeFile.seekg(firstChildIndex, std::ios::beg);
    treeFile.read(reinterpret_cast<char*>(nodeBuffer), sizeof(RTreeNode) * buffSize);
    this->totalSearchIOs += (unsigned int)std::ceil((double)buffSize / nodesPerBlock);
    for (long long i = 0; i < buffSize; i++) {
      if (RTree::intersects(region, nodeBuffer[i].MBR)) {
        this->searchRecursive(region, nodeBuffer[i].firstChildIndex, nodeBuffer[i].lastChildIndex, result);
      }
    }
    delete[] nodeBuffer;
  }
}

File: rtree/src/RTree.java
import algorithms.RTreeAlgorithm;
import structs.RTreeNode;
import structs.Rectangle;

import java.io.*;
import java.util.ArrayList;


/**
 * Clase que representa un RTree
 */
public class RTree {
    private RTreeAlgorithm algorithm;
    private int totalRectBytesRead = 0;
    private int totalNodeBytesRead = 0;
    private int blockSize;
    private RandomAccessFile rectIO;
    private RandomAccessFile nodeIO;

    private DataInputStream rectInputStream;
    private DataInputStream nodeInputStream;

    /**
     * Constructor de la clase RTree
     * @param algorithm Algoritmo de construcción de RTree
     * @param rectanglesPerBlock Cantidad de rectángulos por bloque de disco
     * @param nodesPerBlock Cantidad de nodos por bloque de disco
     */
    public RTree(RTreeAlgorithm algorithm, int diskBlockSize) {
        this.algorithm = algorithm;
        this.blockSize = diskBlockSize;
    }

    public int getTotalBytesRead() {
        return totalNodeBytesRead + totalRectBytesRead;
    }

    public void resetTotalBytesRead() {
        totalRectBytesRead = 0;
        totalNodeBytesRead = 0;
    }
    public int getTotalIOs() {
        int nodeIos = (int) Math.ceil((double) totalNodeBytesRead / (double) blockSize);
        int rectIos = (int) Math.ceil((double) totalRectBytesRead / (double) blockSize);
        return nodeIos + rectIos;
    }
    /**
     * Construye el RTree a partir de un archivo de rectángulos
     * @param filename Nombre del archivo de rectángulos
     * @throws Exception Si ocurre un error al leer el archivo
     */
    public void buildTreeFromFile(String filename) throws Exception {
        this.algorithm.buildTree(filename);
        RandomAccessFile leavesFile = this.algorithm.getLeavesFile();
        RandomAccessFile treeFile = this.algorithm.getTreeFile();
        this.rectIO = leavesFile;
        this.rectInputStream = new DataInputStream(new FileInputStream(leavesFile.getFD()));
        this.nodeIO = treeFile;
        this.nodeInputStream = new DataInputStream(new FileInputStream(treeFile.getFD()));
    }

    /**
     * Busca los rectángulos que intersectan con el rectángulo query
     * @param query Área sobre la cual se realiza la búsqueda
     * @return Lista de rectángulos que intersectan con query
     * @throws Exception Si ocurre un error al leer el archivo
     */
    public ArrayList<Rectangle> search(Rectangle query) throws Exception {
        ArrayList<Rectangle> result = new ArrayList<>();
        this.searchRecursive(query, 0, 0, result);
        return result;
    }

    /**
     * Realiza la búsqueda recursiva de los rectángulos que intersectan con el rectángulo query
     * @param query Área sobre la cual se realiza la búsqueda
     * @param firstChildIndex Índice del primer hijo del nodo actual
     * @param lastChildIndex Índice del último hijo del nodo actual
     * @param result Lista de rectángulos que intersectan con query
     * @throws Exception Si ocurre un error al leer el archivo
     */
    private void searchRecursive(Rectangle query, long firstChildIndex, long lastChildIndex, ArrayList<Rectangle> result) throws Exception {
        // Si los hijos son hojas
        if (firstChildIndex < 0 || lastChildIndex < 0) {
            int rectangleByteSize = Rectangle.SIZE;
            long first = -firstChildIndex / rectangleByteSize;
            long last = -lastChildIndex / rectangleByteSize;
            long buffSize = last - first + 1;
            this.rectIO.seek(-firstChildIndex);
            for (int i = 0; i < buffSize; i++) {
                Rectangle rect = Rectangle.deserialize(rectInputStream);
                if (query.intersects(rect) || rect.intersects(query)) {
                    result.add(rect);
                }
            }
            this.totalRectBytesRead += Math.max(-lastChildIndex - -firstChildIndex, Rectangle.SIZE);
        } else {
            int nodeByteSize = RTreeNode.SIZE;
            long first = firstChildIndex / nodeByteSize;
            long last = lastChildIndex / nodeByteSize;
            long buffSize = last - first + 1;
            this.nodeIO.seek(firstChildIndex);
            for (int i = 0; i < buffSize ; i++) {
                RTreeNode node = RTreeNode.deserialize(nodeInputStream);
                if (query.intersects(node.getMBR()) || node.getMBR().intersects(query)) {
                    this.searchRecursive(query, node.getFirstChildIndex(), node.getLastChildIndex(), result);
                }
            }
            this.totalNodeBytesRead += Math.max(lastChildIndex - firstChildIndex, RTreeNode.SIZE);
        }
    }

    /**
     * Cierra los archivos abiertos
     * @throws IOException Si ocurre un error al cerrar los archivos
     */
    public void cleanup() throws IOException {
        this.algorithm.cleanup();
        this.nodeIO.close();
        this.rectIO.close();
        this.nodeInputStream.close();
        this.rectInputStream.close();
    }
}
